File: examples/android/MSDKSampleAppJava/lib/src/main/java/msdk/java/handlers/Messages.java
package msdk.java.handlers;


import androidx.annotation.NonNull;

import com.evernym.sdk.vcx.VcxException;
import com.evernym.sdk.vcx.utils.UtilsApi;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

import java9.util.concurrent.CompletableFuture;
import msdk.java.logger.Logger;
import msdk.java.messages.Message;
import msdk.java.types.MessageStatusType;
import msdk.java.types.MessageType;
import msdk.java.types.UpdateMessageStatusBody;

/**
 * Class containing methods to work with messages.
 */
public class Messages {
    public static final String TAG = "ConnectMeVcx";

    private Messages() {
    }

    /**
     * Retrieve all pending messages.
     *
     * @return List of {@link Message}
     */
    public static @NonNull
    CompletableFuture<List<Message>> getAllPendingMessages() {
        Logger.getInstance().i("Retrieving pending messages");
        CompletableFuture<List<Message>> result = new CompletableFuture<>();
        try {
            UtilsApi.vcxGetMessages(MessageStatusType.PENDING, null, null).whenComplete((messagesString, err) -> {
                if (err != null) {
                    Logger.getInstance().e("Failed to retrieve messages: ", err);
                    result.completeExceptionally(err);
                    return;
                }
                try {
                    List<Message> messages = new ArrayList<>();
                    JSONArray messagesJson = new JSONArray(messagesString);
                    System.out.println(messagesJson + "getAllPendingMessages");

                    for (int i = 0; i < messagesJson.length(); i++) {
                        JSONArray msgsJson = messagesJson.getJSONObject(i).optJSONArray("msgs");
                        String pairwiseDID = messagesJson.getJSONObject(i).getString("pairwiseDID");
                        if (msgsJson != null) {
                            for (int j = 0; j < msgsJson.length(); j++) {
                                JSONObject message = msgsJson.getJSONObject(j);
                                JSONObject payload = new JSONObject(message.getString("decryptedPayload"));
                                String type = payload.getJSONObject("@type").getString("name");

                                String messageUid = message.getString("uid");
                                String msg = payload.getString("@msg");
                                String status = message.getString("statusCode");
                                messages.add(new Message(pairwiseDID, messageUid, msg, type, status));
                            }
                        }
                    }
                    result.complete(messages);
                } catch (JSONException ex) {
                    result.completeExceptionally(ex);
                }
            });
        } catch (VcxException ex) {
            ex.printStackTrace();
        }
        return result;
    }

    /*
    * Download and find message from the specific thread
    * */
    public static CompletableFuture<Message> downloadNextMessageFromTheThread(
            MessageType messageType,
            String threadId
    ) {
        CompletableFuture<Message> result = new CompletableFuture<>();
        Messages.getAllPendingMessages().handle((messages, throwable) -> {
            if (throwable != null) {
                throwable.printStackTrace();
                result.completeExceptionally(throwable);
            }
            Message foundMessage = null;
            for (Message message : messages) {
                try {
                    String msg = message.getPayload();
                    String pwDid = message.getPwDid();
                    JSONObject msgPayload = new JSONObject(msg);
                    String type = msgPayload.getString("@type");
                    JSONObject thread = msgPayload.getJSONObject("~thread");
                    String thid = thread.getString("thid");

                    if (messageType.equals(MessageType.CREDENTIAL) && messageType.matchesValue(type) && thid.equals(threadId)) {
                        foundMessage = message;
                        break;
                    }
                    if (messageType.equals(MessageType.CONNECTION_RESPONSE) && messageType.matchesValue(type) && pwDid.equals(threadId)) {
                        foundMessage = message;
                        break;
                    }
                    if (messageType.equals(MessageType.ACK) && messageType.matchesValue(type)) {
                        foundMessage = message;
                        break;
                    }
                    if (messageType.equals(MessageType.HANDSHAKE) && messageType.matchesValue(type) && thid.equals(threadId)) {
                        foundMessage = message;
                        break;
                    }
                } catch (JSONException e) {
                    e.printStackTrace();
                }
            }
            result.complete(foundMessage);
            return null;
        });
        return result;
    }

    /*
    * Update status of the message as reviewed
    * */
    public static void updateMessageStatus(String pwDid, String messageId) {
        CompletableFuture<String> result = new CompletableFuture<>();
        try {
            String messagesToUpdate =
                    new UpdateMessageStatusBody()
                        .addMessage(pwDid, messageId)
                        .toJSON();
            UtilsApi.vcxUpdateMessages(MessageStatusType.REVIEWED, messagesToUpdate).whenComplete((v1, error) -> {
                if (error != null) {
                    Logger.getInstance().e("Failed to update messages", error);
                    result.completeExceptionally(error);
                } else {
                    result.complete(null);
                }
            });
        } catch (Exception ex) {
            result.completeExceptionally(ex);
        }
    }
}

File: examples/android/MSDKSampleAppKotlin/app/src/main/java/msdk/kotlin/sample/handlers/Messages.kt
package msdk.kotlin.sample.handlers

import com.evernym.sdk.vcx.VcxException
import com.evernym.sdk.vcx.utils.UtilsApi
import java9.util.concurrent.CompletableFuture
import msdk.kotlin.sample.logger.Logger
import msdk.kotlin.sample.messages.Message
import msdk.kotlin.sample.types.MessageStatusType
import msdk.kotlin.sample.types.MessageType
import msdk.kotlin.sample.types.UpdateMessageStatusBody
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.util.*

/**
 * Class containing methods to work with messages.
 */
object Messages {
    /**
     * Retrieve all pending messages.
     *
     * @return List of [Message]
     */
    val allPendingMessages: CompletableFuture<List<Message>>
        get() {
            Logger.instance.i("Retrieving pending messages")
            val result =
                CompletableFuture<List<Message>>()
            try {
                UtilsApi.vcxGetMessages(MessageStatusType.PENDING, null, null)
                    .whenComplete { messagesString: String?, err: Throwable? ->
                        if (err != null) {
                            Logger.instance
                                .e("Failed to retrieve messages: ", err)
                            result.completeExceptionally(err)
                            return@whenComplete
                        }
                        try {
                            val messages: MutableList<Message> =
                                ArrayList()
                            val messagesJson = JSONArray(messagesString)
                            println(messagesJson.toString() + "getAllPendingMessages")
                            for (i in 0 until messagesJson.length()) {
                                val msgsJson =
                                    messagesJson.getJSONObject(i).optJSONArray("msgs")
                                val pairwiseDID =
                                    messagesJson.getJSONObject(i).getString("pairwiseDID")
                                if (msgsJson != null) {
                                    for (j in 0 until msgsJson.length()) {
                                        val message = msgsJson.getJSONObject(j)
                                        val payload =
                                            JSONObject(message.getString("decryptedPayload"))
                                        val type =
                                            payload.getJSONObject("@type").getString("name")
                                        val messageUid = message.getString("uid")
                                        val msg = payload.getString("@msg")
                                        val status = message.getString("statusCode")
                                        messages.add(
                                            Message(
                                                pairwiseDID,
                                                messageUid,
                                                msg,
                                                type,
                                                status
                                            )
                                        )
                                    }
                                }
                            }
                            result.complete(messages)
                        } catch (ex: JSONException) {
                            result.completeExceptionally(ex)
                        }
                    }
            } catch (ex: VcxException) {
                ex.printStackTrace()
            }
            return result
        }

    /*
    * Download and find message from the specific thread
    * */
    fun downloadNextMessageFromTheThread(
        messageType: MessageType,
        id: String
    ): CompletableFuture<Message?> {
        val result =
            CompletableFuture<Message?>()
        allPendingMessages
            .handle<Any?> { messages: List<Message>, throwable: Throwable? ->
                if (throwable != null) {
                    throwable.printStackTrace()
                    result.completeExceptionally(throwable)
                }
                var foundMessage: Message? = null
                for (message in messages) {
                    try {
                        val msg: String = message.payload
                        val pwDid: String = message.pwDid
                        val msgPayload = JSONObject(msg)
                        val type = msgPayload.getString("@type")
                        val thread = msgPayload.getJSONObject("~thread")
                        val thid = thread.getString("thid")

                        if (messageType == MessageType.CREDENTIAL && messageType.matchesValue(type) && thid == id) {
                            foundMessage = message
                            break
                        }
                        if (messageType == MessageType.CONNECTION_RESPONSE && messageType.matchesValue(type) && pwDid == id) {
                            foundMessage = message
                            break
                        }
                        if (messageType == MessageType.ACK && messageType.matchesValue(type)) {
                            foundMessage = message
                            break
                        }
                        if (messageType == MessageType.HANDSHAKE && messageType.matchesValue(type) && thid == id) {
                            foundMessage = message
                            break
                        }
                    } catch (e: JSONException) {
                        e.printStackTrace()
                    }
                }
                result.complete(foundMessage)
                null
            }
        return result
    }

    /*
    * Update status of the message as reviewed
    * */
    fun updateMessageStatus(pwDid: String?, messageId: String?) {
        val result =
            CompletableFuture<String?>()
        try {
            val messagesToUpdate: String = UpdateMessageStatusBody()
                .addMessage(pwDid!!, messageId!!)
                .toJSON()
            UtilsApi.vcxUpdateMessages(MessageStatusType.REVIEWED, messagesToUpdate)
                .whenComplete { v1: Void?, error: Throwable? ->
                    if (error != null) {
                        Logger.instance.e("Failed to update messages", error)
                        result.completeExceptionally(error)
                    } else {
                        result.complete(null)
                    }
                }
        } catch (ex: Exception) {
            result.completeExceptionally(ex)
        }
    }
}